[java-micronaut] make parent with a discriminator abstract and sealed

auke- opened the pull request:
Parent types with a discriminator which aren't part of the mapped types
should be abstract, since they can't be instantiated themself.

Furthermore an option is added which marks the parent class as sealed in
order to let the Java compiler know which subtypes exist.

<!-- Enter details of the change here. Include additional tests that have been done, reference to the issue for tracking, etc. -->

<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@andriy-dmytruk

andriy-dmytruk commented:
@auke- Thanks for the PR.

I looked at the OpenAPI specification documentation to be sure about what we want to do.

It seems like the mapping property is actually optional and can or can not specify mapping instead of the default one. So, it is probably bad to rely on the default one. Instead looking at the `oneOf` property might be more reasonable. So something like:
```xml
    Book:
      title: Book
      description: Object containg all the info about a book
      type: object
      discriminator:
        propertyName: bookType
      oneOf: [{$ref: '#/components/schemas/Encyclopedia'}, {$ref: '#/components/schemas/Book'}]
```

But this looks very counter-intuitive, and I am no longer sure what is the right solution. Maybe we can just always make class `abstract` as you initially proposed. Adding an option to undo the behavior could be sufficient.

Please, reply if you have any thoughts regarding this, but I will look more to it in the following days.

The OpenAPI documentation says the following (in the [Discriminator Object section](https://spec.openapis.org/oas/latest.html#discriminator-object)).

### Regarding the `mapping` property

> In scenarios where the value of the discriminator field does not match the schema name or implicit mapping is not possible, an optional mapping definition MAY be used:
``` xml
MyResponseType:
  oneOf:
  - $ref: '#/components/schemas/Cat'
  - $ref: '#/components/schemas/Dog'
  - $ref: '#/components/schemas/Lizard'
  - $ref: 'https://gigantic-server.com/schemas/Monster/schema.json'
  discriminator:
    propertyName: petType
    mapping:
      dog: '#/components/schemas/Dog'
      monster: 'https://gigantic-server.com/schemas/Monster/schema.json'
```
> Here the discriminator value of dog will map to the schema #/components/schemas/Dog, rather than the default (implicit) value of Dog. If the discriminator value does not match an implicit or explicit mapping, no schema can be determined and validation SHOULD fail. Mapping keys MUST be string values, but tooling MAY convert response values to strings for comparison.

### Regarding the `oneOf` and `allOf`

  > The discriminator object is legal only when using one of the composite keywords oneOf, anyOf, allOf.
  
  So, it is required to use either `oneOf` on the parent object, or `allOf` on the child object when user wants to achieve polymorphism (in other words, when `discriminator` is used).

> In OAS 3.0, a response payload MAY be described to be exactly one of any number of types:
  ```xml
    MyResponseType:
      oneOf:
      - $ref: '#/components/schemas/Cat'
      - $ref: '#/components/schemas/Dog'
      - $ref: '#/components/schemas/Lizard'
  ```
> which means the payload MUST, by validation, match exactly one of the schemas described by Cat, Dog, or Lizard.

Also:
> In both the oneOf and anyOf use cases, all possible schemas MUST be listed explicitly. To avoid redundancy, the discriminator MAY be added to a parent schema definition, and all schemas comprising the parent schema in an allOf construct may be used as an alternate schema.

auke- commented:
> @auke- Thanks for the PR.
> 
> I looked at the OpenAPI specification documentation to be sure about what we want to do.
> 
> It seems like the mapping property is actually optional and can or can not specify mapping instead of the default one. So, it is probably bad to rely on the default one. Instead looking at the `oneOf` property might be more reasonable. So something like:
> 
> ```
>     Book:
>       title: Book
>       description: Object containg all the info about a book
>       type: object
>       discriminator:
>         propertyName: bookType
>       oneOf: [{$ref: '#/components/schemas/Encyclopedia'}, {$ref: '#/components/schemas/Book'}]
> ```
> 

I used this example with my current implementation. This makes `Book` a concrete class, which made me assume that `codegenModel.discriminator.getMappedModels()` is not just the mapping defined under `mapping` but also the various `oneOf` references. So it doesn't matter how the mapping is defined, either as an explicit `mapping` or implicitly using `oneOf`, the `MappedModel` instances define the actual mapping. In the end this is already used in the templates when generating the `@JsonTypeInfo` and `@JsonSubTypes` annotations.

auke- commented:
I've been fiddling with the various sample specs (e.g. [allOf_composition_discriminator.yaml](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/3_0/allOf_composition_discriminator.yaml)) which does reveal some issues:
- when oneOf is used on the parent type without a mapping everything works as expected
- when allOf is used on the child types (or any descendant) without a mapping the parent is missing (in both the `@JsonSubTypes` and the visitor)
- when oneOf is used with a mapping only the explicitly mapped types are included (this is the case with the example `A`/`B`/`C` where `B` is mapped as `b`)
- when the mapping contains multiple mappings for the same type duplicate visitor methods are generated

When the configuration property `legacyDiscriminatorBehavior` is set to `false` it is even worse since some of the type information is duplicated to the child types but the parent type is still missing.

I'm not sure how to proceed. My use case is covered by this change, but a lot of other use cases aren't...

auke- commented:
I've rebased the change on the latest changes in master, made the behavior configurable and changed the implementation in order to prevent duplicate methods in the visitor when the parent type is already present in the mapping.

This doesn't fix the other issues, but at least makes it usable for my use cases and still supports the sample spec (Animal/Doc/Cat/etc).

andriy-dmytruk commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaMicronautAbstractCodegen.java:
```diff
@@ -138,6 +141,7 @@ public JavaMicronautAbstractCodegen() {
         cliOptions.add(new CliOption(OPT_APPLICATION_NAME, "Micronaut application name (Defaults to the " + CodegenConstants.ARTIFACT_ID + " value)").defaultValue(appName));
         cliOptions.add(CliOption.newBoolean(USE_BEANVALIDATION, "Use BeanValidation API annotations", useBeanValidation));
         cliOptions.add(CliOption.newBoolean(USE_OPTIONAL, "Use Optional container for optional parameters", useOptional));
+        cliOptions.add(CliOption.newBoolean(OPT_USE_SEALED, "Use abstract sealed classes for subtypes with a discriminator", useSealed));
```
Can you add `"(Experimental) "` to the beginning of the description, as the way it works may change.
